drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1195 @@
+					strerror(errno));
+			continue;
+		}
+
+		/* check if the plane can be used by our CRTC */
+		if (plane->possible_crtcs & (1 << out->crtc_index)) {
+			drmModeObjectPropertiesPtr props =
+				drmModeObjectGetProperties(fd, plane_id, DRM_MODE_OBJECT_PLANE);
+
+			/* Get the "type" property to check if this is a primary
+			 * plane. Type property is special, as its enum value is
+			 * defined in UAPI headers. For the properties that are
+			 * not defined in the UAPI headers, we would have to
+			 * give kernel the property name and it would return the
+			 * corresponding enum value. We could also do this for
+			 * the "type" property, but it would make this simple
+			 * example more complex. The reason why defining enum
+			 * values for kernel properties in UAPI headers is
+			 * deprecated is that string names are easier to both
+			 * (userspace and kernel) make unique and keep
+			 * consistent between drivers and kernel versions. But
+			 * in order to not break userspace, some properties were
+			 * left in the UAPI headers as well.
+			 */
+			if (get_property_value(fd, props, "type") == DRM_PLANE_TYPE_PRIMARY) {
+				found_primary = true;
+				out->plane_id = plane_id;
+				ret = 0;
+			}
+
+			drmModeFreeObjectProperties(props);
+		}
+
+		drmModeFreePlane(plane);
+	}
+
+	drmModeFreePlaneResources(plane_res);
+
+	if (found_primary)
+		fprintf(stdout, "found primary plane, id: %d\n", out->plane_id);
+	else
+		fprintf(stdout, "couldn't find a primary plane\n");
+	return ret;
+}
+
+/*
+ * modeset_drm_object_fini() is a new helper function that destroys CRTCs,
+ * connectors and planes
+ */
+
+static void modeset_drm_object_fini(struct drm_object *obj)
+{
+	for (int i = 0; i < obj->props->count_props; i++)
+		drmModeFreeProperty(obj->props_info[i]);
+	free(obj->props_info);
+	drmModeFreeObjectProperties(obj->props);
+}
+
+/*
+ * modeset_setup_objects() is a new function. It helps us to retrieve
+ * connector, CRTC and plane objects properties from the device. These
+ * properties will help us during the atomic modesetting commit, so we save
+ * them in our struct modeset_output object.
+ */
+
+static int modeset_setup_objects(int fd, struct modeset_output *out)
+{
+	struct drm_object *connector = &out->connector;
+	struct drm_object *crtc = &out->crtc;
+	struct drm_object *plane = &out->plane;
+	uint32_t plane_id = out->plane_id;
+	uint32_t crtc_id = out->crtc_id;
+	uint32_t conn_id = out->conn_id;
+	int i;
+
+	/* retrieve connector properties from the device */
+	connector->props = drmModeObjectGetProperties(fd, conn_id, DRM_MODE_OBJECT_CONNECTOR);
+	if (!connector->props) {
+		fprintf(stderr, "cannot get connector %d properties: %s\n", conn_id, strerror(errno));
+		goto out_conn;
+	}
+	connector->props_info = calloc(connector->props->count_props, sizeof(connector->props_info));
+	for (i = 0; i < connector->props->count_props; i++)
+		connector->props_info[i] = drmModeGetProperty(fd, connector->props->props[i]);
```

Comment:
If this chunk was a function with arguments fd, `struct drm_object *`, object ID and object type, you could re-use it rather than just copy it three times.

Reply:
Done.